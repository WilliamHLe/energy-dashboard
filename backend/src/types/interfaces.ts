import mongoose from 'mongoose';
import { IBuilding } from '../api/models/buildings.model';
import { ICategory } from '../api/models/categories.model';

export interface IBuildingCategory {
  category: ICategory,
  buildings: string[]
}

export interface IBuildingEnergyTotal {
  _id: mongoose.Types.ObjectId,
  building: IBuilding,
  total: number,
}

export interface IBuildingScore {
  building: IBuilding,
  score: number,
}

export interface IEnergyCategory {
  category: ICategory,
  total: number
}

export interface ICarrier {
  name: string,
  amount: number
}

export interface IWeeklySaved {
  week: string,
  percentSaved: number,
}

export interface IWeeklyUsage {
  week: string,
  sum: number,
}

export interface ICarrierCategory {
  category: ICategory,
  carriers: ICarrier[]
}

export interface IUsage {
  usage: number,
  date: string
}

export interface IEnergyUsageCategory {
  category: ICategory,
  usage: IUsage[]
}
export interface IEnergyAverage {
  average: number,
  date: string
}

export interface IEnergyAverageByCategory {
  category: ICategory,
  average: IEnergyAverage[]
}

export interface IMetrics {
  energyUsedCurrentYear: number,
  energyUsedLastYear: number,
  area: number,
  buildings: number,
}

export interface IMetricsOut {
  energyUsed: number,
  energySaved: number,
  area: number,
  buildings: number
}

export interface IEnergyUsed {
  category: string,
  energyUsedLastYear: number,
  energyUsedCurrentYear: number,
}
export interface IEnergySaved {
  category: ICategory,
  saved: number,
}
export interface IParsedDates {
  fromDate: Date | undefined
  toDate: Date | undefined
}
