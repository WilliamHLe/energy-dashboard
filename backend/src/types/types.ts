/* eslint-disable camelcase */
/* using camel case for url variables */

export type ReqQueryDate = {
    from_date?: string,
    to_date?: string,
};

export type ReqQuerySearch = {
    name: string
};
