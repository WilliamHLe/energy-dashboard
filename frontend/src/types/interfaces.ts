export interface ICategory {
    id: string,
    name: string
}

export interface IUsageData {
    date: string,
    value: number,
}

interface IUsageAllData {
    category: ICategory,
    usage: IUsageData[],
}

export interface IUsageAll {
    data: IUsageAllData[]
}

export interface IUsageSlug {
    data: IUsageData[]
}

export interface IUsageResponseData {
    x: number,
    y: number,
}

export interface IUsageReturn {
    name: string,
    data: IUsageResponseData[]
}

export interface ICarriers {
    name: string,
    amount: number,
}

interface ICarriersAllData {
    category: ICategory,
    carriers: ICarriers[]
}

export interface ICarriersAll {
    data: ICarriersAllData[]
}

export interface ICarriersAllReturn {
    keys: string[],
    data: any[],
    name: string,
    type: string,
}

export interface ICarriersCategory {
    data: ICarriers[]
}

export interface ICarriersCategoryReturnData {
    name: string,
    y: number,
}

export interface ICarriersCategoryReturn {
    name: string,
    colorByPoint:boolean,
    data: ICarriersCategoryReturnData[]
}

interface ISavedAllData {
    category: ICategory,
    saved: number,
}

export interface ISavedAll {
    data: ISavedAllData[]
}

export interface ISavedAllReturn {
    name: string,
    y: number
}

export interface ISavedTotalData {
    percentSaved: number
}

export interface ISavedTotal {
    data: ISavedTotalData
}

interface ISavedWeeklyData {
    week: string,
    percentSaved: number
}

export interface ISavedWeekly {
    data: ISavedWeeklyData[]
}

export interface IAverageData {
    averageEnergy: number
}

export interface IAverage {
    data: IAverageData
}

export interface IBuildingsData {
    id: string,
    name: string,
    year: number | null,
    area?: number,
    categoryIdEsave?: number,
    categoryDescription?: string,
    category: ICategory,
    tek?: string,
    energyLabel?: string
}

export interface IBuilding {
    data: IBuildingsData
}

export interface IBuildings {
    data: IBuildingsData[]
}

export interface IHighscoreData {
    building: IBuildingsData,
    score: number
}

export interface IHighscore {
    data: IHighscoreData[]
}

export interface IMetricsData {
    energyUsed: number,
    energySaved: number,
    area: number,
    buildings: number,
}

export interface IMetrics {
    data: IMetricsData
}
