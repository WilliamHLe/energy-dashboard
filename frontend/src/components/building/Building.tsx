import React, { useEffect, useState } from 'react';
import axios from 'axios';
import { useParams } from 'react-router';
import style from './building.module.css';
import CategoryUsage from '../buildingCategoryOverview/graphs/CategoryUsage';
import AverageUsage from './graphs/AverageUsage';
import Compare from './compare/Compare';
import HeatMapChart from './heatmap/HeatmapChart';
import Modal from './compare/Modal/Modal';
import EnergySaved from './graphs/EnergySaved';
import b1 from './icons/1.svg';
import b2 from './icons/2.svg';
import b3 from './icons/3.svg';
// s
function Building() {
  interface Ibuilding {
    name: string,
    tek: string,
    area: number,
    year: number,
    energyLabel: string,
    category: {
      id: string,
      name: string,
    }
  }
  // ss
  const { id } = useParams<{ category: string, id: string }>();
  const [modalIsOpen, setModalIsOpen] = useState(false);
  const [compareWithBuilding, setcompareWithBuilding] = useState<Ibuilding>();
  const [currentBuilding, setCurrentBuilding] = useState<Ibuilding>();
  useEffect(() => {
    const fetchCurrentBuildingData = async () => {
      const responseBuilding = await axios.get(`/search?name=${id}`);
      console.log(responseBuilding.data[0]);
      setCurrentBuilding(responseBuilding.data[0]);
    };
    fetchCurrentBuildingData();
  }, [id]);

  const openModal = (building: Ibuilding) => {
    setModalIsOpen(!modalIsOpen);
    setcompareWithBuilding(building);
  };
  console.log(currentBuilding);
  return (
    <div>
      <div>
        <div className={style.building} style={{ opacity: modalIsOpen ? 0.1 : 1.0 }}>
          <div className={`container ${style.energyUsageGraph}`}>
            <CategoryUsage />
          </div>
          <div className={`container ${style.compareBuildings}`}>
            {currentBuilding ? (
              <Compare currentBuilding={currentBuilding} onChange={openModal} />
            ) : <div />}
          </div>
          <div className={`container ${style.previousYears}`}>
            <HeatMapChart />
          </div>
          <div className={`container ${style.energyAverage}`}>
            <AverageUsage />
          </div>
          <div className={`container ${style.energyTotal}`} />
          <div className={`container ${style.getTips}`} />
          <div className={`container ${style.energyPercentageSaved}`}>
            <EnergySaved />
          </div>
          <div className={`container ${style.badges}`}>
            <table style={{ margin: 'auto' }}>
              <tr>
                <td><img alt="Badge" src={b1} /></td>
                <td><img alt="Badge" src={b1} /></td>
                <td><img alt="Badge" src={b1} /></td>
                <td><img alt="Badge" src={b3} /></td>
              </tr>
              <tr>
                <td><img alt="Badge" src={b2} /></td>
                <td><img alt="Badge" src={b2} /></td>
                <td><img alt="Badge" src={b2} /></td>
                <td><img alt="Badge" src={b2} /></td>
              </tr>
            </table>
          </div>
        </div>
        <div>
          {modalIsOpen === true ? (
          // eslint-disable-next-line max-len
            <Modal currentBuilding={currentBuilding} compareBuilding={compareWithBuilding} onChange={openModal} />
          ) : null}
        </div>
      </div>
    </div>
  );
}

export default Building;
