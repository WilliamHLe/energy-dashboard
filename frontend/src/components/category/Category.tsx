import React from 'react';
import { useParams } from 'react-router';
import style from './category.module.css';
import CategotyUsage from './graphs/CategoryUsage';

function Category() {
  const { category } = useParams<{category:string}>();

  return (
    <div className={style.category}>
      <div className={`container ${style.energyUsageGraph}`}>
        <CategotyUsage />
      </div>
      <div className={`container ${style.buildingList}`}>
        Kategori:
        {' '}
        {category}
      </div>
      <div className={`container ${style.energyCarriers}`} />
      <div className={`container ${style.tips}`} />
    </div>
  );
}

export default Category;
