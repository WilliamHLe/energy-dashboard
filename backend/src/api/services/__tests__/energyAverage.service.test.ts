import mongoose from 'mongoose';
import testUtil from '../../../util/__tests__/util';
import energyAverageService from '../energyAverage.service';
import Sensor, { ISensor } from '../../models/sensors.model';
import Building, { IBuilding } from '../../models/buildings.model';
import Category, { ICategory } from '../../models/categories.model';
import { IEnergyAverageByCategory } from '../../../types/interfaces';

interface MockData {
  category: ICategory | null,
  building: IBuilding | null,
  sensor: ISensor | null,
}

const data: MockData = {
  category: null,
  building: null,
  sensor: null,
};

/**
 * Connect to a new in-memory database before running any tests.
 */
beforeAll(async () => {
  await testUtil.connectDatabase();

  data.category = await Category.create({
    _id: new mongoose.Types.ObjectId('6093a8e7c165f7a41d6a613b'), // id hardcoded for snapshot test
    name: 'skole',
  });

  data.building = await Building.create({
    name: 'Hello World',
    year: 2019,
    area: 200,
    categoryIdEsave: 124,
    categoryDescription: 'Text goes here',
    category: data.category.id,
  });

  data.sensor = await Sensor.create({
    description: 'Some description here',
    measurement: 'Fjernvarme',
    name: 'Name of sensor goes here',
    unitOfMeasurement: 'kWh',
    type: 'Forbruksmåler',
    building: data.building.id,
    measurements: [
      {
        date: new Date(2020, 4, 7),
        measurement: 1,
      },
      {
        date: new Date(2020, 0, 3),
        measurement: 2,
      },
      {
        date: new Date(2015, 9, 14),
        measurement: 7,
      },
    ],
  });

  const category = await Category.create({
    _id: new mongoose.Types.ObjectId('6093a8e7c165f7a41d6a6141'), // id hardcoded for snapshot test
    name: 'idrettsbygg',
  });

  const building = await Building.create({
    name: 'Hello World',
    year: 2019,
    area: 200,
    categoryIdEsave: 124,
    categoryDescription: 'Text goes here',
    category: category.id,
  });

  await Sensor.create({
    description: 'Some description here',
    measurement: 'Fjernvarme',
    name: 'Sensor for building not used',
    unitOfMeasurement: 'kWh',
    type: 'Forbruksmåler',
    building: building.id,
    measurements: [
      {
        date: new Date(2020, 4, 7),
        measurement: 200,
      },
      {
        date: new Date(2015, 9, 14),
        measurement: 100,
      },
    ],
  });
});

/**
 * Remove and close the db and server.
 */
afterAll(async () => {
  await testUtil.closeDatabase();
});

describe('averageByYearForBuildings', () => {
  it('Calculates correct yearly average for a list of buildings', async () => {
    const result: number = await energyAverageService.averageByYearForBuildings(
      [data.building?.id],
    );
    // 2kWh in 2020, 7kWh in 2015 => 10kWh total / 2 years => 5 kwH average
    expect(result).toBe(5);
  });
});

describe('averageByYearForCategories', () => {
  it('', async () => {
    const res: IEnergyAverageByCategory[] = await energyAverageService.averageByYearForCategories();
    expect(res).toMatchSnapshot();
  });
});
