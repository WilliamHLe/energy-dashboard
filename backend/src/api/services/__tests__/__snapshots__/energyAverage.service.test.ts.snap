// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`averageByYearForCategories  1`] = `
Array [
  Object {
    "average": 150,
    "category": Object {
      "__v": 0,
      "_id": "6093a8e7c165f7a41d6a6141",
      "name": "idrettsbygg",
    },
  },
  Object {
    "average": 5,
    "category": Object {
      "__v": 0,
      "_id": "6093a8e7c165f7a41d6a613b",
      "name": "skole",
    },
  },
]
`;
